ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PingPacketData.java

```diff
@@ -56,48 +59,62 @@ private PingPacketData(
     this.enrSeq = enrSeq;
   }
 
-  public static PingPacketData create(final Endpoint from, final Endpoint to, final UInt64 enrSeq) {
+  public static PingPacketData create(
+      final Optional<Endpoint> from, final Endpoint to, final UInt64 enrSeq) {
     checkArgument(
         enrSeq != null && UInt64.ZERO.compareTo(enrSeq) < 0, "enrSeq cannot be null or negative");
     return create(from, to, PacketData.defaultExpiration(), enrSeq);
   }
 
   static PingPacketData create(
-      final Endpoint from, final Endpoint to, final long expirationSec, final UInt64 enrSeq) {
+      final Optional<Endpoint> from,
+      final Endpoint to,
+      final long expirationSec,
+      final UInt64 enrSeq) {
     checkArgument(
         enrSeq != null && UInt64.ZERO.compareTo(enrSeq) < 0, "enrSeq cannot be null or negative");
-    return new PingPacketData(Optional.of(from), to, expirationSec, enrSeq);
+    return new PingPacketData(from, to, expirationSec, enrSeq);
   }
 
   public static PingPacketData readFrom(final RLPInput in) {
     in.enterList();
     // The first element signifies the "version", but this value is ignored as of EIP-8
     in.readBigIntegerScalar();
-    final Optional<Endpoint> from = Endpoint.maybeDecodeStandalone(in);
-    final Endpoint to = Endpoint.decodeStandalone(in);
+    Optional<Endpoint> from = Optional.empty();
+    Optional<Endpoint> to = Optional.empty();
+    if (in.nextIsList()) {
+      to = Endpoint.maybeDecodeStandalone(in);
+      // https://github.com/ethereum/devp2p/blob/master/discv4.md#ping-packet-0x01
+      if (in.nextIsList()) { // if there are two, the first is the from address, next is the to
+        // address
+        from = to;
+        to = Endpoint.maybeDecodeStandalone(in);
+      }
+    } else {
+      throw new DevP2PException("missing address in ping packet");
+    }
     final long expiration = in.readLongScalar();
     UInt64 enrSeq = null;
     if (!in.isEndOfCurrentList()) {
       try {
         enrSeq = UInt64.valueOf(in.readLongScalar());
+        LOG.debug("read PING enr as long scalar");
       } catch (MalformedRLPInputException malformed) {
+        LOG.debug("failed to read PING enr as scalar, trying to read bytes instead");
         enrSeq = UInt64.fromBytes(in.readBytes());
       }
     }
     in.leaveListLenient();
-    return new PingPacketData(from, to, expiration, enrSeq);
+    return new PingPacketData(from, to.get(), expiration, enrSeq);
```

Comment:
silly question.  do we care if there is a `to` field on a decoded ping packet?  I don't see the `to` field getting used for anything outside of tests, string serialization, and and `writeTo`.    `writeTo` implies we are sending it rather than receiving it...  So for a decoded packet, is `to` useless?  Do we repeat decoded ping messages?

If we are receiving this packet, maybe we should relax the to requirement and use `.orElse()` or perhaps change the `to` member to Optional<Endpoint> maybeTo and allow malformed ping packet decoding where there are no endpoints at all. 🤔

Reply:
Thats a very good question. You would think the `to` field is far more optional than the `from` field. If we are ignoring what is in the PING in favor of address values at lower layers, then all address fields should be treated as optional. I'll raise this with ACD.